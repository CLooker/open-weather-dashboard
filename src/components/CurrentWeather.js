import React, { Component } from 'react';
import Loading from './Loading';
import apiKey from '../api';
import PropTypes from 'prop-types';

export default class CurrentWeather extends Component {
  static propTypes = {
    setName: PropTypes.func.isRequired,
    units: PropTypes.string.isRequired
  };

  state = {
    name: null,
    temp: null,
    iconCode: null,
    weather: null
  };

  returnUnitsString = units => (units === 'F' ? 'imperial' : 'metric');

  componentWillReceiveProps({ units }) {
    units !== this.props.units && this.fetchData(units);
  }

  // place on the prototype
  // so we can spy that it's invoked
  // by componentDidMount
  fetchData(units) {
    fetch(
      `http://api.openweathermap.org/data/2.5/weather?q=Rome&units=${this.returnUnitsString(
        units
      )}&APPID=${apiKey}`
    )
      .then(res => res.json())
      .then(res =>
        this.setState({
          name: res.name,
          temp: res.main.temp,
          iconCode: res.weather[0].icon,
          weather: res.weather[0].main,
          wind: res.wind.speed,
          clouds: res.clouds.all,
          pressure: res.main.pressure
        })
      )
      .then(() => this.props.setName(this.state.name))
      .catch(err => console.log(err));
  }

  componentDidMount() {
    this.fetchData(this.props.units);
  }

  render() {
    const {
      name,
      weather,
      temp,
      wind,
      iconCode,
      clouds,
      pressure
    } = this.state;

    return (
      <div>
        {name === null ? (
          <Loading />
        ) : (
          <div className="current-weather-container">
            <div className="current-weather">
              <h3> Weather in {name} </h3>
              <div className="weather" style={{ display: 'flex' }}>
                <div className="weather-and-icon">
                  <img
                    src={`http://openweathermap.org/img/w/${iconCode}.png`}
                    alt="Icon depicting current weather"
                  />
                  <h3>{weather}</h3>
                </div>
                <div>
                  <img
                    src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQYVYTeX-IT8VF_M7yQQsClU2CQEJTXgoi8T9jaGhB66jpLWQSw"
                    alt="temperature"
                  />
                  <h3>
                    {temp.toFixed(1)} &deg;{this.props.units}
                  </h3>
                </div>
                <div>
                  <img
                    src={`https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTLHQKHJiUm02mysUrvH8vaKDY-myNAYfRHCUFJgA-LSiNeOKn-`}
                    alt="wind"
                  />
                  <h3>{wind} mph</h3>
                </div>
                <div>
                  {' '}
                  <img
                    src={`http://openweathermap.org/img/w/03d.png`}
                    alt="cloudiness"
                  />
                  <h3>{clouds}%</h3>
                </div>
                <div>
                  <img
                    src={`https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSe31uwyEa-r1R6ntqu4QPim6J6QQsIPqIrZ1RyEYebzC5esPGc`}
                    alt="barometric pressure"
                  />
                  <h3>{pressure} hpa</h3>
                </div>
              </div>
            </div>
          </div>
        )}
      </div>
    );
  }
}
