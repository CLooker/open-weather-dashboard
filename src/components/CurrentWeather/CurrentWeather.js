import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import thermometerIcon from '../../assets/thermometer.png';
import windIcon from '../../assets/wind.png';
import cloudIcon from '../../assets/cloud.png';
import barometerIcon from '../../assets/barometer.png';
import { apiKey, handleConversion, farenheit, celsius } from '../../utils';
import { IconAndText, Loading } from '../common';

const CurrentWeatherWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  border: 1px solid lightgrey;
  border-radius: 10px;
  text-align: center;
`;

export default class CurrentWeather extends PureComponent {
  static propTypes = {
    units: PropTypes.string.isRequired
  };

  state = {
    units: this.props.units,
    temp: '',
    iconCode: '',
    weather: ''
  };

  componentDidMount() {
    const handleResponse = res =>
      this.setState({
        temp: res.main.temp,
        iconCode: res.weather[0].icon,
        weather: res.weather[0].main,
        wind: res.wind.speed,
        clouds: res.clouds.all,
        pressure: res.main.pressure
      });

    fetch(
      `https://api.openweathermap.org/data/2.5/weather?q=Rome&units=imperial&APPID=${apiKey}`
    )
      .then(res => res.json())
      .then(handleResponse)
      .catch(console.error);
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevProps.units !== this.props.units) {
      this.setState({
        units: this.props.units,
        temp: handleConversion(prevState.temp, prevState.units)
      });
    }
  }

  getIsLoading = () => Object.values(this.state).some(value => value === '');

  getCurrentWeatherComponents = () => {
    const {
      weather,
      temp,
      wind,
      iconCode,
      clouds,
      pressure,
      units
    } = this.state;

    const iconAndTextPropsArr = [
      {
        titleAttr: 'Current Weather',
        iconSrc: `https://openweathermap.org/img/w/${iconCode}.png`,
        text: <h4>{weather}</h4>
      },
      {
        titleAttr: 'Current Temperature',
        iconSrc: thermometerIcon,
        text: (
          <h4>
            {temp.toFixed(1)} {units === 'F' ? farenheit : celsius}
          </h4>
        )
      },
      {
        titleAttr: 'Current Wind',
        iconSrc: windIcon,
        text: <h4>{wind} mph</h4>
      },
      {
        titleAttr: 'Current Cloudiness',
        iconSrc: cloudIcon,
        text: <h4>{clouds}%</h4>
      },
      {
        titleAttr: 'Current Barometric Pressure',
        iconSrc: barometerIcon,
        text: <h4>{pressure} hpa</h4>
      }
    ];

    const Wrapper = styled.div`
      display: flex;
    `;

    const IconAndTextWrapper = styled(IconAndText)`
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      padding: 0 40px;
    `;

    return (
      <Wrapper>
        {iconAndTextPropsArr.map(iconAndTextProps => (
          <IconAndTextWrapper
            key={iconAndTextProps.titleAttr}
            {...iconAndTextProps}
          />
        ))}
      </Wrapper>
    );
  };

  render() {
    if (this.getIsLoading()) return <Loading />;

    return (
      <CurrentWeatherWrapper>
        <div>
          <h3>Current Weather</h3>
          {this.getCurrentWeatherComponents()}
        </div>
      </CurrentWeatherWrapper>
    );
  }
}
