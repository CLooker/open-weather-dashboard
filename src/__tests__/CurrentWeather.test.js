import React from 'react';
import { mount, shallow } from 'enzyme';
import CurrentWeather from '../components/CurrentWeather';
import Loading from '../components/Loading';

// refactor fetchData to make it testable

describe('CurrentWeather', () => {
  const props = {
    setName: jest.fn(),
    units: 'F'
  };

  let component, mountedComponent;

  beforeEach(() => {
    component = shallow(<CurrentWeather {...props} />);
    mountedComponent = mount(<CurrentWeather {...props} />);
  });

  describe('initialization', () => {
    it('should render correctly', () => {
      expect(component.debug()).toMatchSnapshot();
    });

    it('should have certain props', () => {
      const actual = component.instance().props;
      const expected = props;
      expect(actual).toEqual(props);
    });

    it('should have certain state', () => {
      const actual = component.state();
      const expected = {
        name: null,
        temp: null,
        iconCode: null,
        weather: null
      };
      expect(actual).toEqual(expected);
    });

    describe('`componentDidMount`', () => {
      let componentDidMountSpy, fetchDataSpy, localMountedComponent;

      beforeEach(() => {
        componentDidMountSpy = jest.spyOn(
          CurrentWeather.prototype,
          'componentDidMount'
        );
        fetchDataSpy = jest.spyOn(CurrentWeather.prototype, 'fetchData');
        localMountedComponent = mount(<CurrentWeather {...props} />);
      });

      it('should call `componentDidMount`', () => {
        expect(componentDidMountSpy).toHaveBeenCalled();
      });

      it('should call `fetchData`', () => {
        expect(fetchDataSpy).toHaveBeenCalled();
      });
    });

    describe('rendering', () => {
      describe('rendering based on `state.name`', () => {
        describe('`name` is null', () => {
          it('should return `Loading`', () => {
            mountedComponent.instance().setState({ name: null });
            const actual = mountedComponent.html();
            const expected = shallow(<Loading />).html();
            expect(actual).toEqual(expect.stringContaining(expected));
          });
        });

        describe('`name` is not null', () => {
          test('that certain state props are rendered', () => {
            const data = {
              name: 'Chicago',
              temp: 44,
              iconCode: 'fakeCode',
              weather: 'cloudy',
              wind: 10,
              clouds: 88,
              pressure: 1000
            };
            mountedComponent.instance().setState(data);
            const actual = mountedComponent.html();
            const expected = Object.keys(data).map(key => data[key]);
            expected.forEach(data => expect(actual).toContain(data));
          });
        });
      });
    });
  });

  describe('methods', () => {
    describe('`returnUnitsString`', () => {
      describe('arg is "F"', () => {
        it('should return "imperial"', () => {
          const actual = mountedComponent.instance().returnUnitsString('F');
          const expected = 'imperial';
          expect(actual).toBe(expected);
        });
      });

      describe('arg is "C"', () => {
        it('should return "metric"', () => {
          const actual = mountedComponent.instance().returnUnitsString('C');
          const expected = 'metric';
          expect(actual).toBe(expected);
        });
      });
    });
  });

  describe('componentWillReceiveProps', () => {});
});
