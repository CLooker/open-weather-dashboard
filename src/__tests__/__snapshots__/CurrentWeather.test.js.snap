// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentWeather initialization should render correctly 1`] = `
"<div>
  <Loading />
</div>"
`;
